benchmarks/pnp_stokes_experiment/linear_pnp_ns.cpp

```diff
@@ -0,0 +1,456 @@
+  // _fasp_bsr_matrix = fasp_format_dcsr_dbsr(&_fasp_matrix, dimension);
+  fasp_bdcsr_free(&_fasp_block_matrix);
+   _fasp_block_matrix.brow = 2;
+   _fasp_block_matrix.bcol = 2;
+   _fasp_block_matrix.blocks = (dCSRmat **)calloc(4, sizeof(dCSRmat *));
+  fasp_dcsr_getblk(&_fasp_matrix, pnp_dofs.val,    pnp_dofs.val,    pnp_dofs.row,
+      pnp_dofs.row,    _fasp_block_matrix[0]);
+  fasp_dcsr_getblk(&_fasp_matrix, pnp_dofs.val,    stokes_dofs.val, pnp_dofs.row,
+       stokes_dofs.row, _fasp_block_matrix[1]);
+  fasp_dcsr_getblk(&_fasp_matrix, stokes_dofs.val, pnp_dofs.val,    stokes_dofs.row,
+     pnp_dofs.row,    _fasp_block_matrix[2]);
+  fasp_dcsr_getblk(&_fasp_matrix, stokes_dofs.val, stokes_dofs.val, stokes_dofs.row,
+     stokes_dofs.row, _fasp_block_matrix[3]);
+
+
+  if (_faps_soln_unallocated) {
+    fasp_dvec_alloc(_eigen_vector->size(), &_fasp_soln);
+    _faps_soln_unallocated = false;
+  }
+  EigenVector_to_dvector_block(_eigen_vector, &_fasp_vector) // to do;
+
+  fasp_dvec_set(_fasp_vector.row, &_fasp_soln, 0.0);
+}
+//--------------------------------------
+std::vector<dolfin::Function> Linear_PNP_NS::fasp_solve () {
+  Linear_PNP_NS::setup_fasp_linear_algebra();
+  std::vector<dolfin::Function> solution(Linear_PNP::get_solutions());
+
+  printf("Solving linear system using FASP solver...\n"); fflush(stdout);
+  INT status = fasp_solver_bdcsr_krylov_pnp_stokes(
+    &_fasp_block_matrix,
+    &_fasp_vector,
+    &_fasp_soln,
+    &_itsolver,
+    &_pnpitsolver,
+    &_pnpamg,
+    &_nsitsolver,
+    &_nsamg,
+    velocity_dofs.row,
+    pressure_dofs.row);
+
+  if (status < 0) {
+    printf("\n### WARNING: FASP solver failed! Exit status = %d.\n", status);
+    fflush(stdout);
+  }
+  else {
+    printf("Successfully solved the linear system\n");
+    fflush(stdout);
+
+    dolfin::EigenVector solution_vector(_eigen_vector->mpi_comm(),_eigen_vector->size());
+    double* array = solution_vector.data();
+
+    std::size_t i
+    // for (std::size_t i = 0; i < _fasp_soln.row; ++i) {
+    //   array[i] = _fasp_soln.val[i];
+    for (i=0; i<_pnp_dofs.row; i++)
+        array[_pnp_dofs.val[i]] = _fasp_soln.val[i];
+    for (i=0; i<stokes_dofs.row;i++)
+        array[_stokes_dofs.val[i]] = _fasp_soln.val[_pnp_dofs.row + i] ;
+
+    dolfin::Function update (
+      Linear_PNP::_convert_EigenVector_to_Function(solution_vector)
+    );
+
+    // This need to change, it copies the same vector 3 times.
+    Function dAnion = update[0];
+    Function dCation = update[1];
+    Function dPotential = update[2];
+    Function dU = update[3];
+    Function dPressure = update[4];
+
+    std::vector<std::shared_ptr<const Function>> vvv = {dAnion,dCation,dPotential};
+    dolfin::Function update_pnp(_functions_space[0])
+    assign(update_pnp , vvv);
+
+    *(solution[0].vector()) += *(update_pnp.vector());
+    *(solution[1].vector()) += *(dU.vector());
+    *(solution[2].vector()) += *(dPressure.vector());
```

Comment:
Here is copies the solutions too many times. My questions is: are the DOF in the same order for the function space and the sub function space. Is the following true ?
`dolfin::Function F1(Subspace);
dolfin::Function F2(BigSpace);
F1.vector()[i] = F2.vector[dof[i]]
`